File: backend/gps_speedometer/gps/simulator.py
"""Fake GPS data generator for development without hardware."""

from __future__ import annotations

import math
import queue
import random
import threading
import time
from datetime import datetime, timezone

SEATTLE_LAT = 47.6062
SEATTLE_LON = -122.3321


class GpsSimulator:
    """Generates synthetic NMEA sentences at 10Hz following a drive pattern."""

    def __init__(self, sentence_queue: queue.Queue[str], update_hz: float = 10.0):
        self.queue = sentence_queue
        self._interval = 1.0 / update_hz
        self._thread: threading.Thread | None = None
        self._stop_event = threading.Event()
        self._connected = True

        # Simulated state
        self._lat = SEATTLE_LAT
        self._lon = SEATTLE_LON
        self._heading = 45.0  # NE
        self._speed_knots = 0.0
        self._target_speed_knots = 35.0  # ~40 mph cruising
        self._altitude = 56.0  # meters
        self._satellites = 12
        self._time_offset = 0.0

    @property
    def connected(self) -> bool:
        return self._connected

    def start(self) -> None:
        self._stop_event.clear()
        self._thread = threading.Thread(target=self._run, daemon=True, name="gps-simulator")
        self._thread.start()

    def stop(self) -> None:
        self._stop_event.set()
        if self._thread:
            self._thread.join(timeout=5.0)
            self._thread = None

    def _run(self) -> None:
        while not self._stop_event.is_set():
            self._update_state()
            now = datetime.now(timezone.utc)

            gga = self._make_gga(now)
            rmc = self._make_rmc(now)

            try:
                self.queue.put_nowait(gga)
                self.queue.put_nowait(rmc)
            except queue.Full:
                pass

            self._stop_event.wait(self._interval)

    def _update_state(self) -> None:
        """Simulate vehicle dynamics — gentle acceleration, turning, speed variation."""
        self._time_offset += self._interval

        # Vary target speed in a pattern (stop at red lights, cruise, accelerate)
        cycle = self._time_offset % 60  # 60-second drive cycle
        if cycle < 5:
            self._target_speed_knots = 0.0  # stopped
        elif cycle < 15:
            self._target_speed_knots = 20.0  # accelerating
        elif cycle < 40:
            self._target_speed_knots = 35.0 + random.uniform(-3, 3)  # cruising
        elif cycle < 50:
            self._target_speed_knots = 50.0 + random.uniform(-2, 2)  # faster
        else:
            self._target_speed_knots = 10.0  # decelerating

        # Smooth speed transitions
        speed_diff = self._target_speed_knots - self._speed_knots
        self._speed_knots += speed_diff * 0.05  # gentle acceleration
        self._speed_knots = max(0, self._speed_knots)

        # Gentle heading drift (simulates curves)
        self._heading += math.sin(self._time_offset * 0.1) * 0.5
        self._heading %= 360

        # Move position based on speed and heading
        speed_mps = self._speed_knots * 0.514444
        dlat = speed_mps * math.cos(math.radians(self._heading)) * self._interval / 111320
        dlon = speed_mps * math.sin(math.radians(self._heading)) * self._interval / (
            111320 * math.cos(math.radians(self._lat))
        )
        self._lat += dlat
        self._lon += dlon

        # Slight altitude variation
        self._altitude += random.uniform(-0.1, 0.1)

    def _make_gga(self, now: datetime) -> str:
        """Generate a GGA sentence."""
        t = now.strftime("%H%M%S.%f")[:9]  # HHMMss.ss
        lat_str, lat_dir = self._format_lat(self._lat)
        lon_str, lon_dir = self._format_lon(self._lon)

        fields = [
            "GPGGA", t, lat_str, lat_dir, lon_str, lon_dir,
            "1",  # fix quality: GPS
            f"{self._satellites:02d}",
            "0.9",  # HDOP
            f"{self._altitude:.1f}", "M",  # altitude
            "-17.0", "M",  # geoid separation
            "", "",  # DGPS fields
        ]
        body = ",".join(fields)
        checksum = self._checksum(body)
        return f"${body}*{checksum}"

    def _make_rmc(self, now: datetime) -> str:
        """Generate an RMC sentence."""
        t = now.strftime("%H%M%S.%f")[:9]
        d = now.strftime("%d%m%y")
        lat_str, lat_dir = self._format_lat(self._lat)
        lon_str, lon_dir = self._format_lon(self._lon)

        fields = [
            "GPRMC", t, "A",  # A=active
            lat_str, lat_dir, lon_str, lon_dir,
            f"{self._speed_knots:.1f}",
            f"{self._heading:.1f}",
            d,
            "", "",  # magnetic variation
            "A",  # mode indicator
        ]
        body = ",".join(fields)
        checksum = self._checksum(body)
        return f"${body}*{checksum}"

    @staticmethod
    def _format_lat(lat: float) -> tuple[str, str]:
        d = "N" if lat >= 0 else "S"
        lat = abs(lat)
        deg = int(lat)
        mins = (lat - deg) * 60
        return f"{deg:02d}{mins:07.4f}", d

    @staticmethod
    def _format_lon(lon: float) -> tuple[str, str]:
        d = "E" if lon >= 0 else "W"
        lon = abs(lon)
        deg = int(lon)
        mins = (lon - deg) * 60
        return f"{deg:03d}{mins:07.4f}", d

    @staticmethod
    def _checksum(sentence: str) -> str:
        cs = 0
        for c in sentence:
            cs ^= ord(c)
        return f"{cs:02X}"

File: v2/src-tauri/src/vehicle/simulator.rs
//! Fake GPS provider for development without hardware — ported from v1
//! `gps/simulator.py`. Generates GGA+RMC sentences following a drive cycle and
//! runs them through the real `NmeaParser`, so the whole GPS path is exercised
//! end-to-end. Random jitter is replaced with deterministic sine variation
//! (keeps it reproducible; the wobble is purely cosmetic).

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::mpsc::SyncSender;
use std::sync::Arc;
use std::thread;
use std::time::Duration;

use super::nmea::{NmeaParser, ParseResult};
use super::{
    send_sample, ReceiverStatus, SampleSource, VehicleEvent, VehicleProvider, VehicleSample,
};

const SEATTLE_LAT: f64 = 47.6062;
const SEATTLE_LON: f64 = -122.3321;
const KNOTS_TO_MPS: f64 = 0.514444;

pub struct SimulatorProvider {
    update_hz: f64,
}

impl SimulatorProvider {
    pub fn new() -> Self {
        SimulatorProvider { update_hz: 10.0 }
    }
}

impl Default for SimulatorProvider {
    fn default() -> Self {
        Self::new()
    }
}

struct SimState {
    lat: f64,
    lon: f64,
    heading: f64,
    speed_knots: f64,
    altitude: f64,
    satellites: u32,
    t: f64,
    interval: f64,
}

impl VehicleProvider for SimulatorProvider {
    fn name(&self) -> &'static str {
        "simulator"
    }

    fn spawn(self: Box<Self>, tx: SyncSender<VehicleEvent>, stop: Arc<AtomicBool>) {
        thread::spawn(move || {
            if tx
                .send(VehicleEvent::Health {
                    source: SampleSource::Simulator,
                    status: ReceiverStatus::Fix,
                })
                .is_err()
            {
                return;
            }
            let interval = 1.0 / self.update_hz;
            let mut st = SimState {
                lat: SEATTLE_LAT,
                lon: SEATTLE_LON,
                heading: 45.0,
                speed_knots: 0.0,
                altitude: 56.0,
                satellites: 12,
                t: 0.0,
                interval,
            };
            let mut parser = NmeaParser::new();

            while !stop.load(Ordering::SeqCst) {
                st.update();
                let gga = st.make_gga();
                let rmc = st.make_rmc();
                // Feed both through the parser; RMC completes a fix.
                parser.parse(&gga);
                if let ParseResult::Fix(fix) = parser.parse(&rmc) {
                    let sample = VehicleSample::from_raw(fix, SampleSource::Simulator);
                    if !send_sample(&tx, sample) {
                        break;
                    }
                }
                thread::sleep(Duration::from_secs_f64(interval));
            }
        });
    }
}

impl SimState {
    fn update(&mut self) {
        self.t += self.interval;

        // 60-second drive cycle: stop, accelerate, cruise, faster, decelerate.
        let cycle = self.t % 60.0;
        let target = if cycle < 5.0 {
            0.0
        } else if cycle < 15.0 {
            20.0
        } else if cycle < 40.0 {
            35.0 + 3.0 * (self.t * 0.7).sin()
        } else if cycle < 50.0 {
            50.0 + 2.0 * (self.t * 0.9).sin()
        } else {
            10.0
        };

        // Smooth speed transitions toward the target.
        self.speed_knots += (target - self.speed_knots) * 0.05;
        if self.speed_knots < 0.0 {
            self.speed_knots = 0.0;
        }

        // Gentle heading drift (simulates curves).
        self.heading += (self.t * 0.1).sin() * 0.5;
        self.heading = self.heading.rem_euclid(360.0);

        // Move position based on speed + heading.
        let speed_mps = self.speed_knots * KNOTS_TO_MPS;
        let dlat = speed_mps * self.heading.to_radians().cos() * self.interval / 111_320.0;
        let dlon = speed_mps * self.heading.to_radians().sin() * self.interval
            / (111_320.0 * self.lat.to_radians().cos());
        self.lat += dlat;
        self.lon += dlon;

        // Slight altitude variation.
        self.altitude += 0.1 * (self.t * 1.3).sin();
    }

    fn make_gga(&self) -> String {
        let (lat_s, lat_d) = format_lat(self.lat);
        let (lon_s, lon_d) = format_lon(self.lon);
        let body = format!(
            "GPGGA,000000.00,{lat_s},{lat_d},{lon_s},{lon_d},1,{sats:02},0.9,{alt:.1},M,-17.0,M,,",
            sats = self.satellites,
            alt = self.altitude,
        );
        with_checksum(&body)
    }

    fn make_rmc(&self) -> String {
        let (lat_s, lat_d) = format_lat(self.lat);
        let (lon_s, lon_d) = format_lon(self.lon);
        let body = format!(
            "GPRMC,000000.00,A,{lat_s},{lat_d},{lon_s},{lon_d},{spd:.1},{hdg:.1},010100,,,A",
            spd = self.speed_knots,
            hdg = self.heading,
        );
        with_checksum(&body)
    }
}

fn format_lat(lat: f64) -> (String, char) {
    let dir = if lat >= 0.0 { 'N' } else { 'S' };
    let lat = lat.abs();
    let deg = lat.trunc();
    let mins = (lat - deg) * 60.0;
    (format!("{:02}{:07.4}", deg as i64, mins), dir)
}

fn format_lon(lon: f64) -> (String, char) {
    let dir = if lon >= 0.0 { 'E' } else { 'W' };
    let lon = lon.abs();
    let deg = lon.trunc();
    let mins = (lon - deg) * 60.0;
    (format!("{:03}{:07.4}", deg as i64, mins), dir)
}

fn with_checksum(body: &str) -> String {
    let cs = body.bytes().fold(0u8, |acc, b| acc ^ b);
    format!("${body}*{cs:02X}")
}
